File: src/map_composer.py
"""Compose a TM2020 .Map.Gbx from converted .Item.Gbx chunks.

Wraps the dotnet helper's ``place-objects-on-map`` command. The helper
needs an existing seed map (any valid .Map.Gbx works as a starting point);
we ship a tiny `assets/empty_stadium.Map.Gbx` for that purpose. It also
needs a 32 m grid of Stadium ground blocks under the items so the player
has somewhere to start before reaching our scenery.

JSON schema mirrors vendor/blendermania-addon/utils/Dotnet.py:150
``DotnetPlaceObjectsOnMap`` exactly so we can fall back to using the
addon's own helper builds if ours ever diverges.
"""
from __future__ import annotations

import math
from dataclasses import dataclass
from pathlib import Path

import numpy as np

from chunker import MeshChunk
from dotnet_runner import (
    CMD_PLACE_OBJECTS_ON_MAP,
    DotnetResult,
    run_dotnet_command,
    write_json_config,
)


# Blender -> TM Position/Rotation conventions. EMPIRICALLY VERIFIED via
# user-confirmed `Alps_yaw_pos.Map.Gbx`:
#
#     TM Position = (-Blender_Y, Blender_Z + 8, -Blender_X)
#     TM Rotation.X = +90 deg (per item, unconditional)
#
# The blendermania-addon's documented formula (MapObjects.py:218-220) is
#     TM Position = (Blender_Y, Blender_Z + 8, Blender_X)
#     TM Rotation.X = Blender_Z_yaw - 90 deg
# The negations on TM X and TM Z plus the +90 (vs the addon's -90) cancel
# the X-mirror and 180-deg yaw that FORZA_TO_TRACKMANIA bakes into our
# content. A hand-modeled Blender item would use the addon's unmodified
# formula directly. (Earlier I baked in the addon formula raw and got
# distance-dependent position errors — the cancellation is required for
# our content.)
ITEM_POSITION_Y_LIFT = 8.0
DEFAULT_ITEM_ROTATION_X_RAD = math.radians(90.0)

BLOCK_GRID_M = 32.0
BLOCK_GRID_Y_OFFSET = 9  # cells, per addon

# TM2020 maps are a positive-only coordinate space anchored at the (0,0,0)
# corner — anything at negative X/Z is literally off the map and TM reports
# the item as "missing". FM4 world coords span both signs, so we shift the
# whole track into positive space before composing. Margins keep the track
# clear of the very edge.
POSITIVE_MARGIN_M = 64.0       # X/Z breathing room from the (0,0) corner
GROUND_CLEARANCE_M = 16.0      # lift so the lowest geometry sits above floor


@dataclass
class PlacedItem:
    """One item placement in the composed map.

    `name` MUST be the path of the item relative to ``<userdir>/Items/``,
    including subdirs and the ``.Item.Gbx`` extension. Example:
    ``"Forzamania/Alps/Alps_Tile_n010_p001_00.Item.Gbx"``. TM2020 uses
    this string as the lookup key against its item library — passing just
    the stem produces "missing item" errors at map load.

    See vendor/blendermania-addon/utils/MapObjects.py:202-211 for the
    canonical convention.

    `item_gbx_path` is the absolute path on disk (the dotnet helper uses
    it for the initial item ingestion).
    """
    name: str
    item_gbx_path: Path
    position_xyz: tuple[float, float, float]
    rotation_xyz: tuple[float, float, float] = (0.0, 0.0, 0.0)


def _to_dotnet_vector3(xyz: tuple[float, float, float]) -> dict:
    """Match the C# DotnetVector3 JSON shape."""
    return {"X": float(xyz[0]), "Y": float(xyz[1]), "Z": float(xyz[2])}


def _to_dotnet_int3(xyz: tuple[int, int, int]) -> dict:
    return {"X": int(xyz[0]), "Y": int(xyz[1]), "Z": int(xyz[2])}


def chunk_to_placed_item(
    item_gbx_path: Path,
    items_rel_path: str,
    blender_center_xyz: tuple[float, float, float],
) -> PlacedItem:
    """Lift one converted chunk into a PlacedItem.

    `items_rel_path` is the item's path relative to ``<userdir>/Items/``
    with forward slashes and the ``.Item.Gbx`` extension — e.g.
    ``"Forzamania/Alps/Alps_Tile_n010_p001_00.Item.Gbx"``. This is the
    string TM2020 looks the item up by.

    `blender_center_xyz` is the chunk's bbox centre in RAW Blender world
    coords (the value blender_export emits to the .center.json sidecar
    after applying FORZA_TO_TRACKMANIA but before centering). We apply
    the empirically-verified Blender->TM conversion here so the item
    lands at the correct TM position.
    """
    bx, by, bz = blender_center_xyz
    # Empirically verified: TM Position = (-Blender_Y, Blender_Z + 8, -Blender_X)
    # The negations cancel the X-mirror our FORZA_TO_TRACKMANIA bakes in;
    # see module-level comment on ITEM_POSITION_Y_LIFT for the full story.
    tm_pos = (-by, bz + ITEM_POSITION_Y_LIFT, -bx)
    return PlacedItem(
        name=items_rel_path,
        item_gbx_path=Path(item_gbx_path),
        position_xyz=tm_pos,
        rotation_xyz=(DEFAULT_ITEM_ROTATION_X_RAD, 0.0, 0.0),
    )


def compute_ground_block_grid(
    placed_items: list[PlacedItem],
    block_name: str = "StadiumPlatform",
    margin_cells: int = 2,
) -> list[dict]:
    """Lay a flat grid of Stadium ground blocks under the placed items.

    Returns DotnetBlock dicts (Name, Dir, Position) snapped to the 32 m
    grid. The grid covers the bbox of all placements + ``margin_cells``
    cells of padding on every side so the player has somewhere to drive
    before falling off the world.
    """
    if not placed_items:
        return []

    xs = [p.position_xyz[0] for p in placed_items]
    zs = [p.position_xyz[2] for p in placed_items]
    x_min, x_max = min(xs), max(xs)
    z_min, z_max = min(zs), max(zs)

    cx_min = math.floor(x_min / BLOCK_GRID_M) - margin_cells
    cx_max = math.floor(x_max / BLOCK_GRID_M) + margin_cells
    cz_min = math.floor(z_min / BLOCK_GRID_M) - margin_cells
    cz_max = math.floor(z_max / BLOCK_GRID_M) + margin_cells

    blocks = []
    for cx in range(cx_min, cx_max + 1):
        for cz in range(cz_min, cz_max + 1):
            blocks.append({
                "Name": block_name,
                "Dir": 0,
                "Position": _to_dotnet_int3((cx, BLOCK_GRID_Y_OFFSET, cz)),
            })
    return blocks


def compose_map(
    dotnet_exe: Path,
    seed_map: Path,
    output_map: Path,
    placed_items: list[PlacedItem],
    block_name: str = "StadiumPlatform",
    config_dir: Path | None = None,
    linux_mode: bool = False,
    wine_cmd: list[str] | None = None,
) -> DotnetResult:
    """Build a .Map.Gbx by stamping items + a ground-block grid into seed_map.

    The dotnet helper writes the composed map next to the seed at
    ``<seed>_<MapSuffix>.Map.Gbx`` by default, OR to ``<output_map>`` if we
    set ShouldOverwrite=True with MapPath=<output_map>. We use the explicit-
    output approach so the result lands in TM's Maps/Forzamania/ folder
    directly.
    """
    if config_dir is None:
        config_dir = output_map.parent

    # --- shift the whole track into TM2020's positive-only space --------
    # Find the min corner across all item positions, then translate every
    # item so that corner lands at (+POSITIVE_MARGIN_M, _, +POSITIVE_MARGIN_M)
    # and the lowest point is GROUND_CLEARANCE_M above the floor. Without
    # this, ~58% of a typical FM4 track sits at negative X/Z = off-map =
    # TM reports every one of those items as "missing".
    if placed_items:
        min_x = min(it.position_xyz[0] for it in placed_items)
        min_y = min(it.position_xyz[1] for it in placed_items)
        min_z = min(it.position_xyz[2] for it in placed_items)
        off_x = POSITIVE_MARGIN_M - min_x
        off_y = GROUND_CLEARANCE_M - min_y
        off_z = POSITIVE_MARGIN_M - min_z
    else:
        off_x = off_y = off_z = 0.0

    def _shifted(xyz: tuple[float, float, float]) -> tuple[float, float, float]:
        return (xyz[0] + off_x, xyz[1] + off_y, xyz[2] + off_z)

    items_payload = []
    shifted_items: list[PlacedItem] = []
    for it in placed_items:
        new_pos = _shifted(it.position_xyz)
        shifted_items.append(PlacedItem(
            name=it.name,
            item_gbx_path=it.item_gbx_path,
            position_xyz=new_pos,
            rotation_xyz=it.rotation_xyz,
        ))
        # Empty strings for the enum fields make the dotnet helper bail
        # with "Unable to parse '' as a value of EPhaseOffset" — defaults
        # come from vendor/blendermania-addon/utils/Dotnet.py:29-66.
        items_payload.append({
            "Name": it.name,
            "Path": str(it.item_gbx_path),
            "Position": _to_dotnet_vector3(new_pos),
            "Rotation": _to_dotnet_vector3(it.rotation_xyz),
            "Pivot": _to_dotnet_vector3((0.0, 0.0, 0.0)),
            "AnimPhaseOffset": "None",
            "DifficultyColor": "Default",
            "LightmapQuality": "Normal",
        })

    # Block grid SKIPPED: `StadiumPlatform` is not a valid TM2020 block
    # name (the dotnet helper silently rejects every one — 0 placed). The
    # working `Alps_yaw_pos` was composed with Blocks=[] and looked right,
    # so leaving the ground out is the right call until we know the real
    # block name. See vendor/blendermania-addon/utils/Constants.py:439:
    #     "GBX.NET can't place free blocks yet"
    # so even with the right name, free-form block placement may be
    # limited. The seed map provides a base ground plane regardless.
    blocks_payload: list[dict] = []
    _ = block_name  # parameter kept for backward-compat callers
    _ = shifted_items

    payload = {
        "MapPath": str(output_map),
        "Blocks": blocks_payload,
        "Items": items_payload,
        "ShouldOverwrite": True,
        "MapSuffix": "_modified",
        "CleanBlocks": True,
        "CleanItems": True,
        "Env": "Stadium2020",
    }

    config_path = config_dir / f"{output_map.stem}_compose.json"
    write_json_config(config_path, payload)

    # The helper doesn't create the seed; copy it into place before invoking
    # so the helper opens-and-modifies in situ.
    if seed_map.is_file() and not output_map.is_file():
        output_map.parent.mkdir(parents=True, exist_ok=True)
        output_map.write_bytes(seed_map.read_bytes())

    return run_dotnet_command(
        dotnet_exe,
        CMD_PLACE_OBJECTS_ON_MAP,
        config_path,
        linux_mode=linux_mode,
        wine_cmd=wine_cmd,
    )

File: app/map_composer.nim
## Compose a TM2020 .Map.Gbx from converted .Item.Gbx chunks.
##
## Nim port of `src/map_composer.py`. Drives nadeo-freeporter's `map` subcommand:
## write a JSON config (the `place-objects-on-map` payload) whose `MapPath` is
## the output map; freeporter generates the blank void base from its own embedded
## seed when that path doesn't exist, so we ship no seed and copy nothing.
##
## The valuable, hard-won part is the Blender→TM coordinate convention below —
## EMPIRICALLY VERIFIED, do not "fix" it. The negations + the +90° rotation
## cancel the X-mirror and 180° yaw that FORZA_TO_TRACKMANIA bakes into our
## content; a hand-modeled Blender item would use the addon's raw formula.

import std/[os, math, json]
import subproc

const
  RC_SUCCESS* = 0
  RC_CONFIG_ERROR* = 3

  # Blender → TM Position/Rotation. EMPIRICALLY VERIFIED via the user-confirmed
  # Alps_yaw_pos.Map.Gbx:  TM Position = (-Blender_Y, Blender_Z + 8, -Blender_X),
  # TM Rotation.X = +90° (unconditional). See module doc for why the negations
  # are required for OUR content (they cancel FORZA_TO_TRACKMANIA's X-mirror +
  # 180° yaw) — the addon's raw formula gave distance-dependent position errors.
  ITEM_POSITION_Y_LIFT* = 8.0
  DEFAULT_ITEM_ROTATION_X_RAD* = degToRad(90.0)

  BLOCK_GRID_M* = 32.0
  BLOCK_GRID_Y_OFFSET* = 9   # cells, per addon

  # TM2020 maps are positive-only, anchored at the (0,0,0) corner — anything at
  # negative X/Z is off-map and reported "missing". FM4 world coords span both
  # signs, so compose_map shifts the whole track positive before composing.
  POSITIVE_MARGIN_M* = 64.0
  GROUND_CLEARANCE_M* = 16.0

type
  Vec3f* = tuple[x, y, z: float64]

  ComposeResult* = object
    returncode*: int
    stdout*:     string
    stderr*:     string

  PlacedItem* = object
    ## One item placement. `name` MUST be the item path relative to
    ## <userdir>/Items/ with the .Item.Gbx extension (TM's library lookup key);
    ## `itemGbxPath` is the absolute path on disk.
    name*:        string
    itemGbxPath*: string
    position*:    Vec3f
    rotation*:    Vec3f

  GroundBlock* = object
    name*:     string
    dir*:      int
    position*: tuple[x, y, z: int]

proc ok*(r: ComposeResult): bool = r.returncode == RC_SUCCESS
proc explanation*(r: ComposeResult): string =
  case r.returncode
  of RC_SUCCESS: "ok"
  of RC_CONFIG_ERROR: "config not found or invalid JSON payload"
  else: "rc=" & $r.returncode

proc chunkToPlacedItem*(itemGbxPath, itemsRelPath: string;
                        blenderCenter: Vec3f): PlacedItem =
  ## Lift one converted chunk into a PlacedItem. `blenderCenter` is the chunk's
  ## bbox centre in RAW Blender world coords (post FORZA_TO_TRACKMANIA, pre
  ## centering — the value blender_export emits to the .center.json sidecar).
  let (bx, by, bz) = blenderCenter
  # TM Position = (-Blender_Y, Blender_Z + 8, -Blender_X)
  PlacedItem(
    name: itemsRelPath,
    itemGbxPath: itemGbxPath,
    position: (-by, bz + ITEM_POSITION_Y_LIFT, -bx),
    rotation: (DEFAULT_ITEM_ROTATION_X_RAD, 0.0, 0.0),
  )

proc computeGroundBlockGrid*(placedItems: seq[PlacedItem];
                             blockName = "StadiumPlatform";
                             marginCells = 2): seq[GroundBlock] =
  ## Flat 32 m grid of Stadium ground blocks covering the placements' bbox +
  ## margin. NOTE: dead in compose_map (blocks_payload is []), kept for API
  ## parity — `StadiumPlatform` isn't a valid TM2020 block name.
  if placedItems.len == 0: return @[]
  var xMin, zMin = Inf
  var xMax, zMax = NegInf
  for p in placedItems:
    xMin = min(xMin, p.position.x); xMax = max(xMax, p.position.x)
    zMin = min(zMin, p.position.z); zMax = max(zMax, p.position.z)
  let cxMin = int(floor(xMin / BLOCK_GRID_M)) - marginCells
  let cxMax = int(floor(xMax / BLOCK_GRID_M)) + marginCells
  let czMin = int(floor(zMin / BLOCK_GRID_M)) - marginCells
  let czMax = int(floor(zMax / BLOCK_GRID_M)) + marginCells
  for cx in cxMin .. cxMax:
    for cz in czMin .. czMax:
      result.add(GroundBlock(name: blockName, dir: 0,
                             position: (cx, BLOCK_GRID_Y_OFFSET, cz)))

proc vec3Json(v: Vec3f): JsonNode =
  ## Match the C# DotnetVector3 JSON shape.
  %*{"X": v.x, "Y": v.y, "Z": v.z}

proc buildComposeConfig*(outputMap: string; placedItems: seq[PlacedItem];
                         blockName = "StadiumPlatform";
                         configDir = ""): string =
  ## Shift the track into positive space, build the freeporter `map` payload,
  ## write the JSON config, and return its path. (Split out of composeMap so it
  ## can be exercised without spawning freeporter.)
  let cfgDir = if configDir.len > 0: configDir else: parentDir(outputMap)

  var offX, offY, offZ = 0.0
  if placedItems.len > 0:
    var minX, minY, minZ = Inf
    for it in placedItems:
      minX = min(minX, it.position.x)
      minY = min(minY, it.position.y)
      minZ = min(minZ, it.position.z)
    offX = POSITIVE_MARGIN_M - minX
    offY = GROUND_CLEARANCE_M - minY
    offZ = POSITIVE_MARGIN_M - minZ

  var itemsPayload = newJArray()
  for it in placedItems:
    let newPos: Vec3f = (it.position.x + offX, it.position.y + offY,
                         it.position.z + offZ)
    itemsPayload.add(%*{
      "Name": it.name,
      "Path": it.itemGbxPath,
      "Position": vec3Json(newPos),
      "Rotation": vec3Json(it.rotation),
      "Pivot": vec3Json((x: 0.0, y: 0.0, z: 0.0)),
      "AnimPhaseOffset": "None",
      "DifficultyColor": "Default",
      "LightmapQuality": "Normal",
    })

  # Block grid SKIPPED: StadiumPlatform isn't a valid TM2020 block name (the
  # helper rejects all of them). blockName kept for backward-compat callers.
  discard blockName
  let payload = %*{
    "MapPath": outputMap,
    "Blocks": newJArray(),
    "Items": itemsPayload,
    "ShouldOverwrite": true,
    "MapSuffix": "_modified",
    "CleanBlocks": true,
    "CleanItems": true,
    "Env": "Stadium2020",
  }

  let (_, stem, _) = splitFile(outputMap)
  let configPath = cfgDir / (stem & "_compose.json")
  createDir(parentDir(configPath))
  writeFile(configPath, pretty(payload, 2))
  return configPath

proc composeMap*(freeporter, outputMap: string; placedItems: seq[PlacedItem];
                 blockName = "StadiumPlatform"; configDir = ""): ComposeResult =
  ## Build a .Map.Gbx by stamping items into a fresh void base map. freeporter
  ## writes the composed map to the config's MapPath, generating the blank base
  ## from its embedded seed when absent — no seed file to ship or copy.
  let configPath = buildComposeConfig(outputMap, placedItems, blockName, configDir)
  createDir(parentDir(outputMap))
  let res = runCaptured(freeporter, @["map", configPath])
  ComposeResult(returncode: res.rc, stdout: res.output, stderr: "")
